### PR TITLE
Add emacs plugin (updated)

### PR DESCRIPTION
**Pull request by oneKelvinSmith**
Updated in response to comments from @bucaran

Port of oh-my-zsh emacs plugin.
Includes a function to kill emacs from command line once
a daemon has started.

More information included in plugins/emacs/README.md.

See reference plugin here:
https://github.com/robbyrussell/oh-my-zsh/tree/master/plugins/emacs

### CODE REVIEW

**Comment by bucaran** [comment]
Thanks! I will be looking at it later today.

File: plugins/emacs/functions/__launch_emacs.fish [hunk 1]
```diff
@@ -0,0 +1,9 @@
+function __launch_emacs
+  set -l x (emacsclient --alternate-editor '' --eval '(x-display-list)' 2>/dev/null)
+
+  if begin; test -z "$x"; or test $x = nil; end
+    emacsclient $argv --alternate-editor '' --create-frame
+  else
+    or emacsclient $argv --alternate-editor ''
```
Comment (bucaran) [inline plugins/emacs/functions/__launch_emacs.fish]: What do you think that `or` is doing there?

File: plugins/emacs/functions/ecd.fish [hunk 2]
```diff
@@ -0,0 +1,12 @@
+function ecd
+  set -l cmd '(let ((buf-name (buffer-file-name (window-buffer))))
+             (if buf-name (file-name-directory buf-name)))'
+  set -l dir (__launch_emacs --eval $cmd | tr -d '\"')
+
+  if test -n "$dir"
+    echo $dir
+  else
+    echo 'cannot deduce current buffer filename.' >/dev/stderr
+    return 1
```
Comment (bucaran) [inline plugins/emacs/functions/ecd.fish]: You don't need this `return 1`.

File: plugins/emacs/functions/__launch_emacs.fish [hunk 3]
```diff
@@ -0,0 +1,9 @@
+function __launch_emacs
+  set -l x (emacsclient --alternate-editor '' --eval '(x-display-list)' 2>/dev/null)
+
+  if begin; test -z "$x"; or test $x = nil; end
```
Comment (bucaran) [inline plugins/emacs/functions/__launch_emacs.fish]: This is more readable:

```fish
if test -z "$x" -o "$x" = nil
...
```
How does `x` becomes `nil`, when?

File: plugins/emacs/emacs.load [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    set -l full_metadata (eval $argv --version)
+    set -l full_version (echo $full_metadata | grep -o "[0-9]\+.[0-9]\+.[0-9]\+.[0-9]\+")
+    set -l major_version (echo $full_version | sed  "s/\..*//")
+  end
+
+  echo $major_version
+end
+
+function __set_editor
+  set -q EDITOR; or set -gx EDITOR emacs
```
Comment (bucaran) [inline plugins/emacs/emacs.load]:
```fish
if not set -q EDITOR
  set -gx EDITOR emacs
end
```

**Comment by bucaran** [comment]
Sorry for the delay and thanks for all the work @oneKelvinSmith

**Comment by bucaran** [comment]
@oneKelvinSmith Comments on the notes?

**Comment by oneKelvinSmith** [comment]
Hey, I'll try work through them tonight :)

**Comment by bucaran** [comment]
Thanks.

File: plugins/emacs/functions/__launch_emacs.fish (hunk 3 above)
Comment (oneKelvinSmith) [inline plugins/emacs/functions/__launch_emacs.fish]: The output of the lisp expression evaluated by emacs is be nil when emacs is not running in a terminal.

**Comment by bucaran** [comment]
@oneKelvinSmith When you have time, open a new PR, some commits got mingled here for some reason. I would also appreciate if you could address each of the comments we discussed previously. :+1:

**Comment by oneKelvinSmith** [comment]
@bucaran 

For this: "I would also appreciate if you could address each of the comments we discussed previously"

Would you like separate commits for each comment, a response on Github per comment or were you thinking of something else?

**Comment by oneKelvinSmith** [comment]
Also, I was making changes on master which was why the other commits got mixed it. I've created a pull request on a new branch.

**Comment by bucaran** [comment]
Thanks!

> Would you like separate commits for each comment, a response on Github per comment or were you thinking of something else?

It's up to you, just make sure to address each of the comments when you can / have time :smile:

**Pull request closed without merging**